docs/site/LB3-vs-LB4-request-response-cycle.md

```diff
@@ -0,0 +1,448 @@
+`final` and `final:after` phases.
+
+The REST API middleware is responsible for creating REST API endpoints out of
+the models in the app. It then uses the configured datasource for connecting
+and querying the undelying database for the corresponding REST requests.
+
+In LoopBack 4, the [sequence](https://loopback.io/doc/en/lb4/Sequence.html) is
+the gatekeeper of all requests to the app. Every request to the app must pass
+through the sequence. It identifies the responsible handler for the requested
+endpoint and passes on the request to the handler.
+
+![LoopBack 4 request/response components](./imgs/lb4-req-res.png)
+
+Unlike LoopBack 3, models in LoopBack 4 do not create REST endpoints. Use the
+`lb4 controller` command to quickly generate the REST endpoints for a model.
+
+For more details, refer to the LoopBack 4
+[request/response cycle](https://loopback.io/doc/en/lb4/Request-response-cycle.html)
+doc.
+
+#### Access to the request/response object
+
+Since LoopBack 3 uses the Express middleware pattern, the request and response
+objects can always be accessed in all the middleware functions.
+
+In LoopBack 4, the request and response objects can be accessed in
+routers loaded using the
+[app.mountExpressRouter() ](https://loopback.io/doc/en/lb4/apidocs.rest.restapplication.mountexpressrouter.html)
+method, using the familiar Express middleware signature.
+
+Controllers, services, and repositories are LoopBack 4 components that
+participate in the request/response cycle. The request and response objects can
+be made available to them via
+[dependency injection](https://loopback.io/doc/en/lb4/Dependency-injection.html)
+.
+
+Example of accesssing the request and response object in a Controller:
+
+```ts
+import {inject} from '@loopback/context';
+import {Request, Response, RestBindings, get} from '@loopback/rest';
+
+export class ExampleController {
+  constructor(
+    @inject(RestBindings.Http.REQUEST) private req: Request,
+    @inject(RestBindings.Http.RESPONSE) private res: Response
+    ) {}
+
+  @get('/headers')
+  headers() {
+    // Sends back the request headers
+    this.res.send(this.req.headers);
+  }
+}
+```
+
+Similarly, the request and response objects can be injected into services
+and respositories along with other objects from the
+[context](https://loopback.io/doc/en/lb4/Context.html).
+
+{% include tip.html content="It may be tempting to use an Express router
+(because of familiarity) instead of a controller to add custom endpoints to the
+app, but bear it in mind that controllers are way more powerful and capable than
+Express routers because of their support for dependency injection and access to
+the application context." %}
+
+#### Data validation
+
+LoopBack 3 models come with
+[validation methods](https://loopback.io/doc/en/lb3/Validating-model-data.html)
+. The validation rules are derived from the model definition files or are
+applied by calling the validation methods in the model's JavaScript file.
```

Comment:
I feel this section is misplaced here. As I understand the content, it describes how LoopBack 4 validates model data at data-access (Repository/ORM) layer. Please note that we are not using AJV there!

Now this page is about REST layer. I agree it's a good idea to describe how validation (and coercion) works at the REST layer, so let's improve/rework the validation section to be more accurate.

In LB3, this part is handled by strong-remoting - see https://github.com/strongloop/strong-remoting/tree/master/lib/types:
- The validation is very minimal, we check only that the provided value matches the expected type (e.g. a string is not allowed for a numeric parameter, but any object will be accepted for object parameters). See `validate` methods in individual types.
- Coercion is pretty comprehensive and applies different rules for values coming from JSON (implemented by `fromTypeValue`) or from string sources like the url query & HTTP headers (implemented by `fromSloppyValue`)

In LB4, we use OpenAPI to describe the type & constraints of request parameters. In a typical controller, we build OpenAPI Schemas describing model data from the metadata provided by LoopBack Models. Non-model parameters (e.g. `id` in `findById(id)`, `filter` in `find(filter)`) are described using inline schema.

At runtime, parameter handling involves several steps:

1. We start with lightweight coercion & validation at OpenAPI parameter level, see https://github.com/strongloop/loopback-next/blob/master/packages/rest/src/coercion/coerce-parameter.ts and https://github.com/strongloop/loopback-next/blob/master/packages/rest/src/coercion/validator.ts
2. Then we convert OpenAPI Schema to JSON Schema and run AJV to validate values against the schema, see https://github.com/strongloop/loopback-next/tree/master/packages/rest/src/validation. I _think_ we are not using AJV-based coercion, but I am not sure. Also at the moment, schema-based validation is implemented for request bodies only, see https://github.com/strongloop/loopback-next/issues/1573

Reply:
@bajtos this section is not about validation at the Repository/ORM layer, it is more about how to do something (validation) in LB4, what you did in LB3 - as closely as possible.

I have updated the content and rephrased some sentences, I hope it makes more sense now.

Reply:
I see. I am afraid I find the new content still misleading. 

Specifically, the following paragraph refers to validation rules like `validatesPresenceOf` that are not applied by the HTTP/REST layer at all. 

```
LoopBack 3 models come with
 [validation methods](https://loopback.io/doc/en/lb3/Validating-model-data.html)
 . The validation rules are derived from the model definition files or are
 applied by calling the validation methods in the model's JavaScript file.
```

I am missing a mention of the validation & coercion rules applied by strong-remoting.

I think it would be really helpful for our users to capture more content & pointer from https://github.com/strongloop/loopback-next/pull/4974/commits/ea52b6f734e19b52dba9ec52b3fc3dea7febd5a0#r408201456 in the docs, this is the kind of information that's almost impossible to find otherwise.

I am proposing the following high-level structure for this section:

1. In LB3, we have strong-remoting validation defined by `accepts` metadata and data-access-level validation defined by model methods like `validatesPresenceOf`. 
2. Briefly describe how strong-remoting validates & coerces data, add links to source code. This way we don't have to go into too much detail now, but still provide useful links to anybody wanting to dig deeper (including us in the future).
3. In LB4, we do full JSON Schema validation at REST layer and property-based validation at data-access level.
4. Briefly describe how LB4 REST layer validates & coerces data, add links to source code.
5. LB4 does not validate non-body parameters against schema yet, + include a link to GH issue.
6. In LB4, it's possible to apply additional validation of model data at REST layer via `jsonSchema` model setting, which can partially replace advanced validations provided by LB3 at data-access level. Explain the caveats.
7. LB4 does not yet offer advanced validation rules at data-access layer like LB3 does, + include a link to GH issue.